### PR TITLE
fix(new-address-modal): new address modal fixes

### PR DESCRIPTION
**Pull request by amovfx**
Added node argument to OpenNewAddressModal
Added nodeName to NewAddressModel
NewAddressModal now uses nodeName from NewAddressModel
Added OpenNewAddressModal to NodeContextMenu

### Description

This is branched off of mint-asset-modal properly.

The modal work flow for generating a taro address is ready for review.

New taro address is updated when genesis info or amount is changed, hitting generate gives a new address

Screenshot 2023-01-13 at 4 44 00 PM

### CODE REVIEW

**Comment by jamaljsr** [comment]
This PR includes the changes from the mint asset PR. You shouldn't squash the commits from that PR as it will make your life hell when that branch changes and you need to update this one. Stacking PRs is hard, and even more cumbersome from a forked repo.  It may be better to wait until the mint asset PR is merged before I can fully review these commits. For now, I'll just do a functional review and provide feedback on the UX side.

File: src/components/common/form/TaroAssetSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    <Form.Item
+      name={name}
+      label={label}
+      help={help}
+      rules={[{ required: true, message: l('cmps.forms.required') }]}
+    >
+      <Select {...rest} onChange={handleChange}>
+        {assets &&
+          assets.map((asset, index) => (
+            <Select.Option key={asset.id} value={index}>
```
Comment (amovfx) [inline src/components/common/form/TaroAssetSelect.tsx:42]: Change this to items.

File: src/components/common/form/TaroNodeSelect.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  const renderComponent = () => {
+    const network = getTaroNetwork(1, 'test network');
+    const bobTaro = createTarodNetworkNode(
+      network,
+      network.nodes.taro[0].version,
+      network.nodes.taro[0].docker,
+      Status.Started,
+    );
+
+    network.nodes.taro.push(bobTaro);
```
Comment (amovfx) [inline src/components/common/form/TaroNodeSelect.spec.tsx:18]: Use add node

File: src/components/designer/NetworkDesigner.tsx [hunk 3]
```diff
@@ -46,6 +47,9 @@ const NetworkDesigner: React.FC<Props> = ({ network, updateStateDelay = 3000 })
   const { zoomIn, zoomOut, zoomReset, ...callbacks } = useStoreActions(s => s.designer);
   const {
     openChannel,
+    mintAsset,
+    sendAsset,
```
Comment (amovfx) [inline src/components/designer/NetworkDesigner.tsx:51]: remove send asset

File: src/components/designer/NetworkDesigner.tsx [hunk 4]
```diff
@@ -94,6 +98,9 @@ const NetworkDesigner: React.FC<Props> = ({ network, updateStateDelay = 3000 })
       {changeBackend.visible && <ChangeBackendModal network={network} />}
       {sendOnChain.visible && <SendOnChainModal network={network} />}
       {advancedOptions.visible && <AdvancedOptionsModal network={network} />}
+      {mintAsset.visible && <MintAssetModal network={network} />}
+      {sendAsset.visible && <SendAssetModal network={network} />}
```
Comment (amovfx) [inline src/components/designer/NetworkDesigner.tsx:102]: remove send asset

File: src/components/designer/taro/actions/OpenSendAssetModal.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    <div onClick={() => showSendAsset()}>
+      <WalletOutlined />
+      <span>{sendLabel}</span>
+    </div>
+  );
+
+  return isContextMenu ? contextMenuButton : sideMenuButton;
+};
+
+export default OpenSendAssetModal;
```
Comment (amovfx) [inline src/components/designer/taro/actions/OpenSendAssetModal.tsx:37]: Remove this

File: src/components/designer/taro/actions/index.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+export { default as MintAssetButton } from './MintAssetButton';
+export { default as OpenSendAssetModal } from './OpenSendAssetModal';
+export { default as MintAssetModal } from './MintAssetModal';
+export { default as SendAssetModal } from './SendAssetModal';
```
Comment (amovfx) [inline src/components/designer/taro/actions/index.ts:4]: Remove send asset modal

File: src/lib/taro/tarod/tarodProxyClient.ts [hunk 7]
```diff
@@ -20,6 +26,18 @@ class TarodProxyClient {
...
+    req: TARO.NewAddressRequest,
+  ): Promise<TARO.NewAddressResponse> {
+    return await this.ipc(ipcChannels.taro.newAddress, { node, req });
+  }
+  async sendAsset(
+    node: TarodNode,
+    req: TARO.SendAssetRequest,
+  ): Promise<TARO.SendAssetResponse> {
+    return await this.ipc(ipcChannels.taro.sendAsset, { node, req });
+  }
```
Comment (amovfx) [inline src/lib/taro/tarod/tarodProxyClient.ts:40]: Maybe remove this

File: src/lib/taro/tarod/tarodService.ts [hunk 8]
```diff
@@ -1,12 +1,50 @@
...
+  ): Promise<PTARO.TaroSendAssetReciept> {
+    const res = await proxy.sendAsset(this.cast(from), req);
+    console.log(JSON.stringify(res.taroTransfer, null, 2));
+    return {
+      transferTxid: res.transferTxid.toString(),
+      anchorOutputIndex: res.anchorOutputIndex,
+      transferTxBytes: res.transferTxBytes.toString(),
+      totalFeeSats: res.totalFeeSats,
+    };
+  }
```
Comment (amovfx) [inline src/lib/taro/tarod/tarodService.ts:39]: Remove

File: src/lib/taro/types.ts [hunk 9]
```diff
@@ -2,8 +2,11 @@
  * Shared types to normalize interfaces between the different
  * Taro implementations
  */
-
-export interface TaroAsset {
+export interface TaroAssetCommon {
+  balance?: string;
+  amount?: string;
+}
+export interface TaroAsset extends TaroAssetCommon {
```
Comment (amovfx) [inline src/lib/taro/types.ts:9]: This makes the type a bit more flexible for components that handle TaroAssets and Tarobalances.

File: src/shared/ipcChannels.ts [hunk 10]
```diff
@@ -22,5 +22,8 @@ export default {
   taro: {
     listAssets: 'taro-list-assets',
     listBalances: 'taro-list-balances',
+    mintAsset: 'taro-mint-asset',
+    newAddress: 'taro-new-address',
+    sendAsset: 'taro-send-asset',
```
Comment (amovfx) [inline src/shared/ipcChannels.ts:27]: remove

File: src/shared/tarodDefaults.ts [hunk 11]
```diff
@@ -16,9 +16,14 @@ export const defaultTarodListBalances = (
...
 });
 
+export const defaultTarodMintAsset = (): MintAssetResponse => ({
+  batchKey: Buffer.from('none'),
+});
+
 const defaults = {
   [ipcChannels.taro.listAssets]: defaultTarodListAssets,
   [ipcChannels.taro.listBalances]: defaultTarodListBalances,
+  [ipcChannels.taro.mintAsset]: defaultTarodMintAsset,
```
Comment (amovfx) [inline src/shared/tarodDefaults.ts:26]: add default for new address

File: src/shared/tarodTypes.ts [hunk 12]
```diff
@@ -3,3 +3,12 @@ export type { ListAssetResponse__Output as ListAssetResponse } from '@hodlone/ta
...
 export type { ListBalancesResponse__Output as ListBalancesResponse } from '@hodlone/taro-api/dist/types/tarorpc/ListBalancesResponse';
+
+export type { MintAssetRequest } from '@hodlone/taro-api/dist/types/tarorpc/MintAssetRequest';
+export type { MintAssetResponse__Output as MintAssetResponse } from '@hodlone/taro-api/dist/types/tarorpc/MintAssetResponse';
+
+export type { NewAddrRequest as NewAddressRequest } from '@hodlone/taro-api/dist/types/tarorpc/NewAddrRequest';
+export type { Addr__Output as NewAddressResponse } from '@hodlone/taro-api/dist/types/tarorpc/Addr';
+
+export type { SendAssetRequest } from '@hodlone/taro-api/dist/types/tarorpc/SendAssetRequest';
+export type { SendAssetResponse__Output as SendAssetResponse } from '@hodlone/taro-api/dist/types/tarorpc/SendAssetResponse';
```
Comment (amovfx) [inline src/shared/tarodTypes.ts:14]: Remove

File: src/store/models/modals.ts [hunk 13]
```diff
@@ -52,6 +52,26 @@ interface AssetInfoModel {
...
+  visible: boolean;
+  nodeName?: string;
+}
+
+interface SendAssetModel {
+  visible: boolean;
+  balanceIndex?: number;
+  to?: string;
+  amount?: number;
+}
```
Comment (amovfx) [inline src/store/models/modals.ts:65]: Remove

File: src/store/models/taro.ts [hunk 14]
```diff
@@ -13,6 +19,41 @@ export interface TaroNodeModel {
...
+  genesisBootstrapInfo: string;
+  amount: string;
+}
+
+export interface SendAssetPayload {
+  from: TaroNode;
+  to: TaroNode;
+  genesisBootstrapInfo: string;
+  amount: number;
+}
```
Comment (amovfx) [inline src/store/models/taro.ts:43]: Remove

File: src/types/index.ts [hunk 15]
```diff
@@ -163,6 +164,15 @@ export interface TaroService {
...
+  mintAsset: (
+    node: TaroNode,
+    req: TARO.MintAssetRequest,
+  ) => Promise<TARO.MintAssetResponse>;
+  newAddress: (node: TaroNode, req: TARO.NewAddressRequest) => Promise<PTARO.TaroAddress>;
+  sendAsset: (
+    from: TaroNode,
+    req: TARO.SendAssetRequest,
+  ) => Promise<PTARO.TaroSendAssetReciept>;
 }
```
Comment (amovfx) [inline src/types/index.ts:176]: remove

File: src/utils/tests/renderWithProviders.tsx [hunk 16]
```diff
@@ -25,6 +25,9 @@ export const taroServiceMock: jest.Mocked<TaroService> = {
   listAssets: jest.fn(),
   listBalances: jest.fn(),
   waitUntilOnline: jest.fn(),
+  mintAsset: jest.fn(),
+  sendAsset: jest.fn(),
```
Comment (amovfx) [inline src/utils/tests/renderWithProviders.tsx:29]: remove

**Review by amovfx — COMMENTED** [review]
Did a review of my own, to double check what could be improved and get the code more inline with your previous review.

**Comment by amovfx** [comment]
I just got your comment of waiting for merges to happen and branch retargeting. Thanks for the advice. The branching I've been doing has been causing me a lot of headaches. I'll slow down.

**Comment by jamaljsr** [comment]
Now that the mint-asset PR has been merged, would you like to reopen this one and rebase it on the latest `feat/add-taro` branch? I can then review this?

**Pull request closed without merging**